### PR TITLE
Move some semantic functions from func.d to funcsem.d and typesem.d

### PR DESCRIPTION
**Pull request by muscar**

### CODE REVIEW

**Comment by muscar** [comment]
Draft for now to see what tests fail and what C++ interfaces need updating.

File: compiler/src/dmd/func.d [hunk 1]
```diff
@@ -1501,27 +1501,6 @@ extern (C++) class FuncDeclaration : Declaration
...
 
-    /****************************************************
-     * Determine whether an 'out' contract is declared inside
-     * the given function or any of its overrides.
-     * Params:
-     *      fd = the function to search
-     * Returns:
-     *      true    found an 'out' contract
-     */
-    static bool needsFensure(FuncDeclaration fd) @safe
```
Comment (thewilsonator) [inline compiler/src/dmd/func.d:1512]: you will need to update `frontend.h` to remove this from that generated header

**Review by thewilsonator — APPROVED** [review]
fails `runnable/test16098.d` and DMD's unit tests.

This otherwise looks good.

File: compiler/src/dmd/semantic3.d [hunk 2]
```diff
@@ -314,7 +314,7 @@ private extern(C++) final class Semantic3Visitor : Visitor
         fds.checkInContractOverrides();
 
         // Remember whether we need to generate an 'out' contract.
-        immutable bool needEnsure = FuncDeclaration.needsFensure(funcdecl);
+        immutable bool needEnsure = needsFensure(funcdecl);
```
Comment (ibuclaw) [inline compiler/src/dmd/semantic3.d:317]: Just a small note that D does have UFCS, and it is not wrong to take advantage of it.

Comment (muscar) [reply compiler/src/dmd/semantic3.d:317]: Good catch. Cheers.

**Comment by muscar** [comment]
I see it fails when running the executable for `runnable/test16098.d` on 32 bit platforms with optimizations on.

@RazvanN7 you reviewed Walter's PRs in which added the test. Do you have any more context around it/any hints?

Thanks.

```
Program received signal SIGSEGV, Segmentation fault.
0xf7e0f14a in gc_qalloc () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#0  0xf7e0f14a in gc_qalloc () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#1  0xf7e364db in _d_criticalenter2 () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#2  0xf7e45657 in rt.trace._staticDtor_L408_C1() () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#3  0xf7e4690b in rt.trace.__moddtor() () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#4  0xf7e40263 in rt.minfo.runModuleFuncsRev!(rt.minfo.ModuleGroup.runTlsDtors().__lambda1).runModuleFuncsRev(const(immutable(object.ModuleInfo)*)[]) () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#5  0xf7e3fa8a in rt.minfo.ModuleGroup.runTlsDtors() () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#6  0xf7e3ff5b in rt.minfo.rt_moduleTlsDtor().__foreachbody1(ref rt.sections_elf_shared.DSO) () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#7  0xf7e41c7b in rt.sections_elf_shared.DSO.opApplyReverse(scope int(ref rt.sections_elf_shared.DSO) delegate) () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#8  0xf7e3ff23 in rt_moduleTlsDtor () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#9  0xf7e37141 in rt_term () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#10 0xf7e37813 in rt.dmain2._d_run_main2(char[][], uint, extern(C) int(char[][]) function).runAll() () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#11 0xf7e3769a in rt.dmain2._d_run_main2(char[][], uint, extern(C) int(char[][]) function).tryExec(scope void() delegate) () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#12 0xf7e375f1 in _d_run_main2 () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#13 0xf7e373f9 in _d_run_main () from /home/runner/work/dmd/dmd/compiler/test/../../../phobos/generated/linux/release/32/libphobos2.so.0.109
#14 0x5655642f in main ()
#15 0xf7621519 in __libc_start_call_main (main=main@entry=0x565563f8 <main>, argc=argc@entry=2, argv=argv@entry=0xffffc744) at ../sysdeps/nptl/libc_start_call_main.h:58
#16 0xf76215f3 in __libc_start_main_impl (main=0x565563f8 <main>, argc=2, argv=0xffffc744, init=0x0, fini=0x0, rtld_fini=0xf7fcaab0 <_dl_fini>, stack_end=0xffffc73c) at ../csu/libc-start.c:392
#17 0x565560bb in _start ()
 ... runnable/test16098.d            -fPIC (-inline -release -g -O)==============================
Test 'runnable/test16098.d' failed. The logged output:
/home/runner/work/dmd/dmd/generated/linux/release/32/dmd -conf= -m32 -Irunnable  -fPIC  -od/home/runner/work/dmd/dmd/compiler/test/test_results/runnable/d -of/home/runner/work/dmd/dmd/compiler/test/test_results/runnable/d/test16098_0  runnable/test16098.d 
/home/runner/work/dmd/dmd/compiler/test/test_results/runnable/d/test16098_0  --DRT-testmode=run-main

/home/runner/work/dmd/dmd/generated/linux/release/32/dmd -conf= -m32 -Irunnable  -fPIC -inline -od/home/runner/work/dmd/dmd/compiler/test/test_results/runnable/d -of/home/runner/work/dmd/dmd/compiler/test/test_results/runnable/d/test16098_1  runnable/test16098.d 
/home/runner/work/dmd/dmd/compiler/test/test_results/runnable/d/test16098_1  --DRT-testmode=run-main

/home/runner/work/dmd/dmd/generated/linux/release/32/dmd -conf= -m32 -Irunnable  -fPIC -release -od/home/runner/work/dmd/dmd/compiler/test/test_results/runnable/d -of/home/runner/work/dmd/dmd/compiler/test/test_results/runnable/d/test16098_2  runnable/test16098.d 
/home/runner/work/dmd/dmd/compiler/test/test_results/runnable/d/test16098_2  --DRT-testmode=run-main
Segmentation fault (core dumped)
```

**Comment by muscar** [comment]
OK, since I don't have access to an x86 linux machine to debug the failure, I've moved `checkNRVO` back to `func.d`. It looks like moving the function in `funcsem.d` breaks `runnable/test16098.d` on x86 linux. It seems likely that the compiler generates wrong code in that case. Probably something around stack alignment, judging by the issue the test is supposed to cover. Since the refactoring was fairly mechanical, it might be that the imports between the two files are different and `checkNRVO` ends up using some slightly different functions?

File: compiler/src/dmd/funcsem.d [hunk 3]
```diff
@@ -2844,3 +2865,176 @@ bool setUnsafePreview(Scope* sc, FeatureState fs, bool gag, Loc loc, const(char)
...
+                // Variables in the data segment (e.g. globals, TLS or not),
+                // parameters and closure variables cannot be NRVOed.
+                if (v.isDataseg() || v.isParameter() || v.toParent2() != fd)
+                    return false;
+                if (v.nestedrefs.length && fd.needsClosure())
+                    return false;
+                // don't know if the return storage is aligned
+                version (MARS)
+                {
+                    if (alignSectionVars && (*alignSectionVars).contains(v))
```
Comment (RazvanN7) [inline compiler/src/dmd/funcsem.d]: Maybe this was your problem? This should be `fd.alignSectionVars`

**Comment by RazvanN7** [comment]
@muscar I suggest you do such changes at a smaller granularity level so that it is easy to debug.

**Pull request merged**